File: files/2026-06-21/CSEC 21062026.R
library(tidyverse)
library(readxl)

path <- "2026-06-21/Challenge-Jun21.xlsx"
input1 <- read_excel(path, range = "B3:D10")
input2 <- read_excel(path, range = "G3:L17")
test <- read_excel(path, range = "E3:E10")

result <- input1 %>%
  left_join(
    input2 %>% pivot_longer(-Grades, names_to = "year", values_to = "value"),
    by = c(`Job Grade` = "Grades")
  ) %>%
  mutate(
    `Range Difference` = coalesce(value - lag(value), 0),
    .by = `Job Grade`
  ) %>%
  filter(Salary <= value) %>%
  slice_head(n = 1, by = `Job Grade`) %>%
  select(`Range Difference`)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE

File: files/2026-06-21/CSEC 21062026.py
import pandas as pd

path = "2026-06-21/Challenge-Jun21.xlsx"
input1 = pd.read_excel(path, usecols="B:D", nrows=8, skiprows=2)
input2 = pd.read_excel(path, usecols="G:L", nrows=15, skiprows=2)
test = pd.read_excel(path, usecols="E", nrows=7, skiprows=2)

input2_long = input2.melt(id_vars="Grades", var_name="year", value_name="value")

result = (
    input1.merge(input2_long, left_on="Job Grade ", right_on="Grades", how="left")
    .assign(
        **{
            "Range Difference": lambda df: df.groupby("Job Grade ")["value"].transform(
                lambda x: x.diff().fillna(0)
            )
        }
    )
    .loc[lambda df: df["Salary"] <= df["value"]]
    .groupby("Job Grade ", sort=False)
    .first()
    .reset_index()[["Range Difference"]]
    .astype("int64")
)

print(result.equals(test))
# True
